Firestore/core/src/model/value_util.cc

```diff
@@ -486,6 +491,105 @@ std::string CanonicalId(const google_firestore_v1_ArrayValue& value) {
+      value->timestamp_value.seconds = std::numeric_limits<int64_t>::min();
+      value->timestamp_value.nanos = 0;
+      return value;
+    }
+
+    case google_firestore_v1_Value_string_value_tag: {
+      Message<google_firestore_v1_Value> value;
+      value->which_value_type = value_tag;
+      value->string_value = nullptr;
+      return value;
+    }
+
+    case google_firestore_v1_Value_bytes_value_tag: {
+      Message<google_firestore_v1_Value> value;
+      value->which_value_type = value_tag;
+      value->bytes_value = nullptr;
+      return value;
+    }
+
+    case google_firestore_v1_Value_reference_value_tag:
+      return RefValue(DatabaseId("", ""), DocumentKey::Empty());
+
+    case google_firestore_v1_Value_geo_point_value_tag: {
+      Message<google_firestore_v1_Value> value;
+      value->which_value_type = value_tag;
+      value->geo_point_value.latitude = -90.0;
+      value->geo_point_value.longitude = -180.0;
+      return value;
+    }
+
+    case google_firestore_v1_Value_array_value_tag: {
+      Message<google_firestore_v1_Value> value;
+      value->which_value_type = value_tag;
+      value->array_value.values = nullptr;
+      value->array_value.values_count = 0;
+      return value;
+    }
+
+    case google_firestore_v1_Value_map_value_tag: {
+      Message<google_firestore_v1_Value> value;
+      value->which_value_type = value_tag;
+      value->map_value.fields = nullptr;
+      value->map_value.fields_count = 0;
+      return value;
+    }
+
+    default:
+      HARD_FAIL("Invalid type value: %s", value_tag);
+  }
+}
+
+Message<google_firestore_v1_Value> GetUpperBound(pb_size_t value_tag) {
+  switch (value_tag) {
+    case google_firestore_v1_Value_null_value_tag:
+      return GetLowerBound(google_protobuf_BoolValue_value_tag);
+    case google_firestore_v1_Value_boolean_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_integer_value_tag);
+    case google_firestore_v1_Value_integer_value_tag:
+    case google_firestore_v1_Value_double_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_timestamp_value_tag);
+    case google_firestore_v1_Value_timestamp_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_string_value_tag);
+    case google_firestore_v1_Value_string_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_bytes_value_tag);
+    case google_firestore_v1_Value_bytes_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_reference_value_tag);
+    case google_firestore_v1_Value_reference_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_geo_point_value_tag);
+    case google_firestore_v1_Value_geo_point_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_array_value_tag);
+    case google_firestore_v1_Value_array_value_tag:
+      return GetLowerBound(google_firestore_v1_Value_map_value_tag);
+    case google_firestore_v1_Value_map_value_tag:
+      return MaxValue();
```

Comment:
FYI We can return the underlying proto here as the memory allocation is static and we don't free it.